### ISSUE TITLE
Improve interaction with eldoc

### ISSUE THREAD

**Issue by memeplex** [company-mode/company-mode#797]
Add company modes commands to `eldoc-message-commands` with:

```
(eldoc-add-command-completions "company-")
```

This is necessary to avoid eldoc messages to be permanently deleted by company manipulations of the buffer.  I've been debugging this with elpy call signature hints and it works much better after the change.

The documentation of  `eldoc-message-commands` states:

> Commands after which it is appropriate to print in the echo area.

You are probably more capacitated than I to exclude any special company command that could potentially message to the echo area, but that's a start.

It's useful to trace `eldoc--message-command-p` in order to find out offenders.

As simple as the above sounds it is hard won knowledge :(

**Comment by dgutov** [comment]
I'm going to copy my comment from a closed issue here:

But `meta` shows information about the symbol you're inserting. ElDoc doesn't really know about it, so it can't show the most pertinent tip.

So in this case I'd rather pick a solution that gives priority to Company over ElDoc.

**Comment by memeplex** [comment]
Your comment is ok for the other issue but this is completely different. Here is not about suppressing the output of company in the echo area but avoiding company commands that show no message at all to suppress the output of every other module. Eldoc uses a whitelist of commands after which is ok to show output. Most (if not all) company comments fall under this category and are silently triggered by innocent looking self inserting commands that are of course whitelisted. Adding company stuff to the whitelist is not to the disadvantage of company but to the advantage of the rest of eldoc users out there. If one company command does output to the echo area, then it shouldn't be in the whitelist, of course; but  to my knowledge no company command falls under this category since the messaging is triggered by a timer, not a command. Notice that all of the above would still be required even if the echo area frontend were disabled o didn't exist at all.

**Comment by memeplex** [comment]
Btw I find the whitelist-based design a weak one as you can read in http://debbugs.gnu.org/cgi/bugreport.cgi?bug=31749 but nevertheless these are the rules to play by.

**Comment by dgutov** [comment]
What you're saying is suppressing ElDoc output in this situation is kinda difficult. I agree. We can get around it, for instance, by advising `eldoc-display-message-no-interference-p`.

So we can disable ElDoc while the popup is displayed. That might work to a disadvantage of those backends that don't support `meta`, however.

**Comment by memeplex** [comment]
I don't quite understand you answer. Are you saying that whitelisting the
company- prefix is suppressing the echo area frontend output? It shouldn't
be so but if it were you only have to remove the specific command which
output was being suppresed from the whitelist.

**Comment by memeplex** [comment]
Just to be sure we're on the same page this is all that I'm proposing (except for also removing some specific command afterwards):

```
(eldoc-add-command-completions "company-")
```

Is that interfering with normal company operation somehow?

**Comment by dgutov** [comment]
> Are you saying that whitelisting the company- prefix is suppressing the echo area frontend output?

I think it will. Or, rather, might make the echo area quickly alternate between the two outputs in an annoying fashion.

**Comment by memeplex** [comment]
> Are you saying that whitelisting the company- prefix is suppressing the
> echo area frontend output?
>
> I think it will. Or, rather, might make the echo area quickly alternate
> between the two outputs in an annoying fashion.
>

Yes but that already happens. Advertising every company command as a
messaging command fails to address the flickering problem. If anything it
increases the message switching rate. And also introduces flickering when
the echo area frontend is disabled!

Now if some command that really outputs to the echo area (I'm unable to
find one) is being muted let's fix that removing it from the whitelist. But
otherwise whitelisting silent commands is the right thing to do, and
tricking eldoc into thinking that everything company does is echoing a
message will not make company messages more permanent but more insistent by
the way of flickering.

I agree a further mechanism is necessary if you want company's messages to
"win" over eldoc while company is displaying its menu. But I still remark
that whitelisting silent commands is playing by the rules, that it won't
cripple company in any way and that at least will give the users the
ability to *effectively* disable the echo area frontend if that's their
wish.

**Comment by memeplex** [comment]
>
>  address the flickering problem. If anything it increases the message
> switching rate.
>

Well the exact effect will depend on the user typing velocity, the major
mode eldoc triggers and the configured eldoc timers.

**Comment by dgutov** [comment]
> Advertising every company command as a messaging command fails to address the flickering problem.

Yes. What I'm saying, we should address this problem in the other direction, and "disable ElDoc while the popup is displayed".

> tricking eldoc into thinking that everything company does is echoing a message will not make company messages more permanent but more insistent by the way of flickering

Indeed. Like I said, a different solution is needed. Like advising `eldoc-display-message-no-interference-p`.

**Comment by memeplex** [comment]
>
> Yes, I've read you and I fully agree but please whitelist company for the
sake of users wanting to disallow the meta frontend and because that is the
entire point of the whitelist: letting eldoc know which commands are silent.

There are two issues here:

1. Company shouldn't suppress other commands output when the meta frontend
is disabled. This is fixed by playing by eldoc rules (whitelisting).

2. Company meta frontend, when enabled, wants to get full control of the
echo area. Here whitelisting has nothing to do, at most it will have an
unpredictable effect based on contingent factors. I agree another solution
is needed.

Since adding company to the whitelist is the right thing to do (TM), fixes
1 and is not predictably related to 2, which anyway will have to be fixed
other way, I don't see any compelling reason not to do it.

>

**Comment by dgutov** [comment]
Since removing the metadata frontend is something a user has to do themselves, they can whitelist the company commands on the next line in their config. What might be helpful is a wiki entry describing this.

On the flip side, the other users could use instructions to disable `eldoc-mode` while the company popup is visible. Probably using `company-completion-started-hook` and `company-after-completion-hook`.

**Comment by dgutov** [comment]
Here's a change I'm think of:

```diff
diff --git a/company.el b/company.el
index 05a599c..d204bfb 100644
--- a/company.el
+++ b/company.el
@@ -646,6 +646,11 @@ return a string prefixed with one space."
   :type '(choice (const :tag "off" nil)
                  (const :tag "on" t)))
 
+;; https://github.com/company-mode/company-mode/issues/797
+(defcustom company-inhibit-eldoc t
+  "If non-nil, suppress ElDoc messages while completion is active."
+  :type 'boolean)
+
 (defvar company-async-wait 0.03
   "Pause between checks to see if the value's been set when turning an
 asynchronous call into synchronous.")
@@ -3208,5 +3213,11 @@ Delay is determined by `company-tooltip-idle-delay'."
     (`post-command (company-echo-show-when-idle 'company-fetch-metadata))
     (`hide (company-echo-hide))))
 
+(defun company--eldoc-allowed-p ()
+  (not (and company-inhibit-eldoc company-candidates (current-message))))
+
+(advice-add #'eldoc-display-message-no-interference-p :before-while
+            #'company--eldoc-allowed-p)
+
 (provide 'company)
 ;;; company.el ends here
```

It should make ElDoc not step on the toes of the meta frontend is it's enabled. And it should be compatible with your preferred configuration as well (meta frontend disabled, company commands triggering eldoc).

BTW, which company commands do you want to trigger eldoc exactly? I can only think of `company-complete`, `company-complete-common` and `company-complete-number`.

**Comment by memeplex** [comment]
For a real-world example, please check [here](https://github.com/jorgenschaefer/elpy/issues/1389) the modifications that were necessary to improve eldoc/company interaction in elpy.

**Comment by memeplex** [comment]
Also check: https://github.com/jorgenschaefer/elpy/search?q=eldoc-add-command-completions&unscoped_q=eldoc-add-command-completions

**Comment by lionel-** [comment]
This didn't work for me:

```elisp
(eldoc-add-command-completions "company-")
```

This seems to work:

```elisp
;; Nuclear option
(defun company-echo-hide ())
(defun company-echo-frontend (command))
(defun company-echo-strip-common-frontend (command))
(defun company-echo-metadata-frontend (command))
```

Is there (or should be) a better way to disable echo messages by company completions?

**Comment by dgutov** [comment]
@lionel- Try simply removing `company-echo-metadata-frontend` from `company-frontends`?
